### DISCUSSION TITLE
Set default conda kernel, or choose more cleverly?

### DISCUSSION THREAD

**Discussion by moble** [quarto-dev/quarto-cli#10716] [Q&A]
### Description

I have a lot of conda kernels that do some fairly weird stuff.  For example, one automatically opens an SSH tunnel to a cluster and starts a very specific version of python there to use as the kernel.  I don't want to use that kernel to build my quarto files (especially because it fails when I'm not connected to the internet), but that's actually the one that quarto selects somehow.  I have a vanilla kernel that I use for most of my work that I'd prefer to use by default, but I can't see how to make that happen.  I've read [the docs](https://quarto.org/docs/computations/python.html#kernel-selection), but don't see anything relevant — even to explain *how* quarto is choosing what it's choosing.

I've also seen [this discussion](https://github.com/quarto-dev/quarto-cli/discussions/8850) about global defaults.  The main reasoning against having such things appears to be

> For us, reproducibility (especially as content moves from being rendered locally to rendered, say, on CI) is a central concern, and ambient user-wide configurations like this tend to cause problems.

But if I want other people to reproduce my document, being able to use a vanilla kernel would actually improve their chances.  On the other hand, having to specify the `jupyter` metadata option makes it less reproducible.

Is there a good way to handle this?  Or is there at least some consistent way quarto is selecting the kernel to use?

**Comment by cderv** [comment]
There is a way with virtual to make Quarto use the python version from an active env. I don't think we have that for Conda, and we should definitely have it by checking active `CONDA_DEFAULT_ENV`. But it should work if conda put the env version first on PATH I think 🤔 

Otherwise, the way to tell Quarto to use a version of Python without looking for it is to set `QUARTO_PYTHON` environment variable. If not set, it will look on PATH or use Python Launcher on windows. It will then take first kernelspec that provide python language.

**Reply by cderv** [reply]
Though reading again your feedback, I understand that the right conda environment may be found, but that you have a lot of kernels that provide python language support and you want to use a specific one as default without having to set `jupyter` configuration. Is that right ? 

Could you share result of `quarto check jupyter` in one of your project ?

**Reply by moble** [reply]
> the right conda environment may be found

Yes, this appears to be the case.


> you have a lot of kernels that provide python language support and you want to use a specific one as default without having to set jupyter configuration

That's right. 


> Could you share result of `quarto check jupyter` in one of your project ?

Due to an error (which is beside the point, since it only happens because it's using this weird kernel), the output is so long that github won't let me include the whole thing, so I'll snip most of the error:

```
❯ quarto check jupyter                                                    
Quarto 1.5.57
[✓] Checking Python 3 installation....OK
      Version: 3.11.9 (Conda)
      Path: /Users/boyle/.continuum/current/envs/science/bin/python3
      Jupyter: 5.7.2
      Kernels: SXS.jl, rik_ssh_black_python, detecting_boosts, python3, julia-1.10, sagemath, julia-1.8, conda-env-science-py, julia-1.9, julia-1.7, spherical, conda-env-jupyterlab-py, sxs_virtualenv, quaternionic, rik_ssh_meistri_python, rik_ssh_cbsu_python, scri, julia, gwframes, wolframlanguage11.3

(|) Checking Jupyter engine render....[I 14:01:42.978 remote_ikernel] Remote kernel version: 0.4.6.
[I 14:01:42.978 remote_ikernel] File location: /Users/boyle/.continuum/current/envs/science/lib/python3.11/site-packages/remote_ikernel/kernel.py.
[I 14:01:42.978 remote_ikernel] Launching kernel over SSH.
[I 14:01:42.978 remote_ikernel] Login command: 'ssh -o StrictHostKeyChecking=no  black'.
(-) Checking Jupyter engine render....[I 14:01:44.860 remote_ikernel] Established connection; starting kernel.
[I 14:01:44.861 remote_ikernel] Current working directory /private/var/folders/bh/rg4bdsl965l0572z3vtmwkr40000gn/T/quarto-sessionaa4a1a8b4c049faf/69d065b96aa71ef7.
(|) Checking Jupyter engine render....[I 14:01:45.020 remote_ikernel] Running kernel command: '/home/boyle/anaconda3/bin/ipython kernel -f ./rik_kernel-96115925-ab51-4602-b1a4-428fc760bb51.json'.
(|) Checking Jupyter engine render....[I 14:02:15.807 remote_ikernel] Setting up tunnels on ports: 63786, 63783, 63784, 63785, 63787.
(-) Checking Jupyter engine render....

An error occurred while executing the following cell:
------------------

[... snip ...]

FileNotFoundError                         Traceback (most recent call last)
<ipython-input-1-9237d2e707ca> in <module>()
    189   # hex-decode the path
    190   run_path = base64.b64decode(run_path.encode("utf-8")).decode("utf-8")
--> 191   os.chdir(run_path)
    192 
    193 # reset state

FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/bh/rg4bdsl965l0572z3vtmwkr40000gn/T/quarto-sessionaa4a1a8b4c049faf/69d065b96aa71ef7'

[I 14:02:24.603 remote_ikernel] Caught interrupt; sending SIGINT to kernel.
[✓] Checking Jupyter engine render....OK
```

Here, quarto is evidently choosing the `rik_ssh_black_python` kernel, whereas I want `python3`.  I guess it just chooses the first kernel in that list that has `"language": "python"` — skipping `SXS.jl` since that's a julia kernel.

(Irrelevant detail:  Judging from what happens when I try to compile a `qmd`, I believe the error is happening because quarto presumably created `run_path` on my laptop, but the kernel is running on a different system, so it obviously can't find `run_path`.  This isn't something I expect quarto to be able to handle; I just want quarto to use a different kernel.)

**Reply by cderv** [reply]
> I guess it just chooses the first kernel in that list that has "language": "python"

Yes that is what happens. It will choose the first, unless `jupyter` config is set at project level or document level. https://quarto.org/docs/computations/python.html#kernel-selection

> whereas I want python3

Are you suggesting will default to any kernel named `python3` first, and if none, then use one that as python language available ? 

I am thinking about what we could do to improve the situation. I agree this is not a good default. 

Regarding conda environment, if you create one specific for your quarto project, and activate it, then the right kernel is used as all the other would not be available. is that what you observe ?

**Reply by moble** [reply]
There is evidently a config option `MultiKernelManager.default_kernel_name`, which defaults to `python3`.  I don't know how to query for that, but if I run `jupyter server --generate-config`, it's in the generated file.

> Regarding conda environment, if you create one specific for your quarto project, and activate it, then the right kernel is used as all the other would not be available. is that what you observe ?

No, the other kernels are still available.  And I think that's right; I believe kernels are supposed to be sort of at a higher level than envs.  For example, kernels on my Mac all exist in `~/Library/Jupyter/kernels`, rather than an env-specific directory.

**Reply by cscheid** [reply]
> And I think that's right; I believe kernels are supposed to be sort of at a higher level than envs

Yes, this is a Jupyter feature that I believe to be a design mistake - the list of Jupyter kernels is stored globally, and there's no way to check the consistency between Jupyter kernels and available Python setups in the environment. This is particularly salient when distributions upgrade Python installations and leave stale Jupyter kernels behind; there's nothing we can do to guard against this possible source of problems.

**Reply by moble** [reply]
Fair enough, but I'll point out that in my case the kernels and python envs *are* consistent, and nothing stale is causing any problems.  I actively use the various kernels.

**Reply by cderv** [reply]
> For example, kernels on my Mac all exist in ~/Library/Jupyter/kernels, rather than an env-specific directory.

This is probably where preferring any kernel with language python located in `CONDA_DEFAULT_ENV` make sense. We did that for virtual env - I'll add for Conda. This will at least solve this use case : Creating a specific conda env for a project, and get default kernel for this env unless specified otherwise.

**Reply by moble** [reply]
> any kernel with language python located in `CONDA_DEFAULT_ENV`

I don't understand this.  I don't think any kernel will ever be located inside an env, and I don't know of any way that the env will know about a particular kernel.  The only connection I see is that the first element of `argv` in the kernel.json ***may*** contain the name of the env — depending on how it's created.

If I do
```
conda create -y -n test_kernel_env python ipykernel
conda activate test_kernel_env
python -m ipykernel install --user --name=my_new_kernel
```
it installs the kernel in `/Users/moble/Library/Jupyter/kernels/my_new_kernel`, and nothing I see in the env changes.

**Reply by cderv** [reply]
> The only connection I see is that the first element of argv in the kernel.json may contain the name of the env — depending on how it's created.

I think this is the assumption we did for virtual env. Quarto will prefer any kernel with supported language that is found in virtual env path. If none, that default behavior is done. So this is a way to prefer the current virtual env installation when it is active.
